### DISCUSSION TITLE
Undo history?

### DISCUSSION THREAD

**Discussion by dusty-phillips** [dexie/Dexie.js#1396] [Ideas]
It occurred to me that the cloud sync protocol might have sufficient information to implement “seamless undo”. One of the few things I like about redux is the easy undo history. I’ve been toying with dexie middleware for undo, but it feels very manual. Any thoughts on weather undo can be had for free with cloud?

**Comment by dfahlander** [comment]
Currently not storing undo information. Quite sparse with the info saved into the changes-tables. I think it is better to have a dedicated middleware for that. Dexie.Observale did store undo info but I think it was a mistake to store now than necessary for sync. One option could be to have it optional and per table somehow.

**Comment by gotjoshua** [comment]
I am working on a middleware these days, that i will share when it's a bit more polished.

I agree that it feels a bit more manual than I'd like, but i do enjoy getting deep into the inners of dexie, as that will help me to use the lib effectively.
I am looking forward to upgrading  to 4.0 as many of the changes in the road map sound amazing.

**Reply by floer32** [reply]
@gotjoshua did you end up making this middleware?

**Reply by gotjoshua** [reply]
> making this middleware?

I have a proof of concept working, but it is quite alpha level stability so far. And i haven't tested it on any other tool chains other than pnpm vite preact chrome.

https://gitlab.com/onezoomin/bygonz/bygonz/-/blob/trunk/packages/Bygonz/src/BygonzMiddleware.ts

I am happy for any and all feedback and collaboration!

**Reply by gotjoshua** [reply]
I focused on human readability of the history data and sync strategies so far, using Dgraph as a very performant backend.

Over view diagram:


![Bygonz_whitebg](https://user-images.githubusercontent.com/8686454/165873628-2144e628-8557-43f7-9d89-26c7a6f1fa8d.png)
